pkg/datapath/iptables/iptables.go

```diff
@@ -1032,7 +1034,19 @@ func (m *IptablesManager) doGetProxyPort(prog iptablesInterface, name string) ui
 		return 0
 	}
 
-	re := regexp.MustCompile(name + ".*TPROXY redirect (0.0.0.0|::):([1-9][0-9]*) mark")
+	// The DNS proxy uses localhost now, as it was changed from 0.0.0.0 in
+	// GH-25309. If any other listener / proxy is used, then fallback to
+	// 0.0.0.0, hence the else clause, in order to prevent regressions where
+	// other parts of Cilium assume 0.0.0.0 is still being used.
+	restr := name + ".*TPROXY redirect "
+	if name == proxytypes.DNSProxyName {
+		restr += "(" + ipfamily.IPv4().Localhost + "|" + ipfamily.IPv6().Localhost + ")"
+	} else {
+		restr += "(" + "0.0.0.0" + "|" + "::" + ")"
```

Comment:
Also Envoy proxy listeners now use the localhost addresses, see https://github.com/cilium/cilium/pull/24011. Then there is the upgrade case, where the iptables rule still has the old any addresses from a previous Cilium version. Hence it would be better to simply add all of these forms as alternatives to the regex, e.g.:
```
  re := regexp.MustCompile(name + ".*TPROXY redirect (0.0.0.0|" + ipfamily.IPv4().Localhost + "|::|" + ipfamily.IPv6().Localhost + "):([1-9][0-9]*) mark")
```
Also maybe keep some of the old forms in the tests to test the upgrade case?

Finally an observation: the IPv4 and IPv6 forms are using the same port number, so it does not matter which one matches.

Reply:
I've used the regex provided. 

For the tests, I believe [this](https://github.com/christarazi/cilium/blob/pr/christarazi/fix-reuse-old-proxy-port/pkg/datapath/iptables/iptables_test.go#L296-L340) case should already cover the upgrade, right?